### PR TITLE
CASMINST-5984 add to management-nodes-rollout docs

### PR DESCRIPTION
**Pull request by leliasen-hpe**
# Description

# Checklist

- [ ] If I added any command snippets, the steps they belong to follow the prompt conventions (see [example][1]).
- [ ] If I added a new directory, I also updated `.github/CODEOWNERS` with the corresponding team in [Cray-HPE][2].
- [ ] My commits or Pull-Request Title contain my JIRA information, or I do not have a JIRA.

 
[1]: https://github.com/Cray-HPE/docs-csm/blob/main/introduction/documentation_conventions.md#using-prompts
[2]: https://github.com/Cray-HPE/teams

### CODE REVIEW

File: operations/iuf/stages/management_nodes_rollout.md [hunk 1]
```diff
@@ -3,17 +3,19 @@
...
+**`NOTE`** `management-nodes-rollout` currently does not rebuild management NCN storage nodes or `ncn-m001`. These nodes can be personalized with the CFS configuration created in the [update-cfs-config](update_cfs_config.md) or they can be manually rebuilt into the new customized image, both options have the same result. 
 
-- [Stage 1 - Ceph image upgrade](#../../upgrade/Stage_1.md) describes the rebuild process for NCN storage nodes
-- [Stage 2.3 - `ncn-m001` upgrade](#../../upgrade/Stage_2.md#stage-23---ncn-m001-upgrade) describes the rebuild process for `ncn-m001`
+- To personalize managment storage nodes or management master nodes, follow the instructions for `management-nodes-rollout` in the [upgrade all products documentation](../workflows/upgrade_all_products.md#652-ncn-master-nodes).
+- To manually rebuild ths NCNs, then proceed to [manually rebuild NCNs with specific image and CFS configuration](#manually-rebuild-ncns-with-specific-image-and-cfs-configuration).
```
Comment (deantroe) [inline operations/iuf/stages/management_nodes_rollout.md:9]:
```suggestion
- To manually rebuild the NCNs, proceed to [manually rebuild NCNs with specific image and CFS configuration](#manually-rebuild-ncns-with-specific-image-and-cfs-configuration).
```

File: operations/iuf/stages/management_nodes_rollout.md [hunk 2]
```diff
@@ -26,22 +28,89 @@ The following arguments are most often used with the `management-nodes-rollout`
...
-(`ncn-m001#`) Execute the `management-nodes-rollout` stage for activity `admin-230127` using the default concurrent management rollout percentage and limiting the operation to `Management_Worker` nodes.
+## Examples
 
+1. (`ncn-m001#`) Execute the `management-nodes-rollout` stage for activity `admin-230127` using the default concurrent management rollout percentage and limiting the operation to `Management_Worker` nodes.
+
+    ```bash
+    iuf -a admin-230127 run --limit-management-rollout Management_Worker -r management-nodes-rollout
+    ```
+
+    Expected Behavior: All NCN worker nodes will rebuilt. Each set of worker nodes that is being rebuilt, should contin 20% of total number of worker nodes. For example, if there are 10 total worker nodes, then 2 should be rebuilt at a time.
```
Comment (deantroe) [inline operations/iuf/stages/management_nodes_rollout.md:53]:
```suggestion
    Expected behavior: All NCN worker nodes will rebuilt. Each set of worker nodes that is being rebuilt will contain 20% of the total number of worker nodes. For example, if there are 10 total worker nodes, then 2 will be rebuilt at a time. 
```

File: operations/iuf/stages/management_nodes_rollout.md [hunk 3]
```diff
@@ -26,22 +28,89 @@ The following arguments are most often used with the `management-nodes-rollout`
...
 
-## Example
+The `cmrp` parameter limits the percentage of worker nodes rolled out in parallel. The worker node rebuild can coordinate rebuilding multiple worker nodes at once. It starts by rebuilding one worker node. Once the node has been removed from the cluster, the workflow checks if it safe to rebuild the next worker node based on what services are running in the cluster. If it is safe, it will proceed in rebuilding the next node (parts of this will be in parallel with the first worker node rebuild). If it is unsafe to rebuild becuase the cluster could get into a bad state, then it waits to rebuild the second node. The `cmrp` parameter selects the percentage of worker nodes that the worker node rebuild should coordinate rebuilding at one time. For example, if there are 15 worker nodes and `cmrp=33`, then 5 worker nodes will be rebuilt in one worker node rebuild workflow. The worker node rebuild workflow will rebuild each of the 5 worker nodes and it may rebuild some in parallel depending on if it is safe to do so.
```
Comment (deantroe) [inline operations/iuf/stages/management_nodes_rollout.md:43]:
```suggestion
The `-cmrp` argument limits the percentage of worker nodes rolled out in parallel. The worker node rebuild can coordinate rebuilding multiple worker nodes at once. It starts by rebuilding one worker node. Once that node has been removed from the system, the workflow checks if it is safe to rebuild the next worker node based on what services are running in the system. If it is safe, it will proceed to rebuild the next node, partially in parallel with the first worker node rebuild. If it is unsafe to rebuild in parallel because the system could get into a bad state, then it waits to rebuild the second node until it is safe. The `-cmrp` parameter selects the percentage of worker nodes that the worker node rebuild should rebuild at one time. For example, if there are 15 worker nodes and `-cmrp 33` is specified, then 5 worker nodes will be rebuilt at once and with as much parallelization as possible given the state of the system.
```

File: operations/iuf/stages/management_nodes_rollout.md [hunk 4]
```diff
@@ -26,22 +28,89 @@ The following arguments are most often used with the `management-nodes-rollout`
...
+    iuf -a admin-230127 run --limit-management-rollout Management_Worker -r management-nodes-rollout
+    ```
+
+    Expected Behavior: All NCN worker nodes will rebuilt. Each set of worker nodes that is being rebuilt, should contin 20% of total number of worker nodes. For example, if there are 10 total worker nodes, then 2 should be rebuilt at a time. 
+
+1. (`ncn-m001#`) Execute the `management-nodes-rollout` stage for activity `admin-230127` using the following parameters.
+
+    - Assume 10 worker nodes (`ncn-w001` through `ncn-w010`)
+    - `limit-management-rollout=[Managment_Worker Management_Master]` 
+    -  `cmrp=35`
```
Comment (deantroe) [inline operations/iuf/stages/management_nodes_rollout.md:59]:
```suggestion
    -  `-cmrp 35`
```

File: operations/iuf/stages/management_nodes_rollout.md [hunk 5]
```diff
@@ -26,22 +28,89 @@ The following arguments are most often used with the `management-nodes-rollout`
...
+    ```bash
+    iuf -a admin-230127 run --limit-management-rollout Management_Worker -r management-nodes-rollout
+    ```
+
+    Expected Behavior: All NCN worker nodes will rebuilt. Each set of worker nodes that is being rebuilt, should contin 20% of total number of worker nodes. For example, if there are 10 total worker nodes, then 2 should be rebuilt at a time. 
+
+1. (`ncn-m001#`) Execute the `management-nodes-rollout` stage for activity `admin-230127` using the following parameters.
+
+    - Assume 10 worker nodes (`ncn-w001` through `ncn-w010`)
+    - `limit-management-rollout=[Managment_Worker Management_Master]`
```
Comment (deantroe) [inline operations/iuf/stages/management_nodes_rollout.md:58]:
```suggestion
    - `-limit-management-rollout [Managment_Worker Management_Master]` 
```

File: operations/iuf/stages/management_nodes_rollout.md [hunk 6]
```diff
@@ -26,22 +28,89 @@ The following arguments are most often used with the `management-nodes-rollout`
...
+
+    - Assume 10 worker nodes (`ncn-w001` through `ncn-w010`)
+    - `limit-management-rollout=[Managment_Worker Management_Master]` 
+    -  `cmrp=35`
+    - `ncn-w004` and `ncn-m002` are labled with  `prevent-iuf-rollout=true` 
+
+    First, label `ncn-w004` and `ncn-m002` with `iuf-prevent-rollout=true`. Then execute the following command.
+
+    ```bash
+    iuf -a admin-230127 run --limit-management-rollout Management_Worker Management_Master  --cmrp 33 -r management-nodes-rollout
```
Comment (deantroe) [inline operations/iuf/stages/management_nodes_rollout.md:65]: Does this format work, or should the 2 roles be comma separated?

File: operations/node_management/Rebuild_NCNs/Rebuild_NCNs.md [hunk 7]
```diff
@@ -66,6 +66,9 @@ make sure that the following conditions are met:
 ```bash
 /usr/share/doc/csm/upgrade/scripts/rebuild/ncn-rebuild-master-nodes.sh ncn-m002
 ```
+> **`NOTES:`**
```
Comment (deantroe) [inline operations/node_management/Rebuild_NCNs/Rebuild_NCNs.md:69]: Why the ">"?

File: operations/node_management/Rebuild_NCNs/Rebuild_NCNs.md [hunk 8]
```diff
@@ -66,6 +66,9 @@ make sure that the following conditions are met:
...
+- This script should be run from ncn-m002 when rebuilding ncn-m001.
```
Comment (deantroe) [inline operations/node_management/Rebuild_NCNs/Rebuild_NCNs.md:70]:
```suggestion
- This script should be run from `ncn-m002` when rebuilding `ncn-m001`.
```

File: operations/node_management/Rebuild_NCNs/Rebuild_NCNs.md [hunk 9]
```diff
@@ -66,6 +66,9 @@ make sure that the following conditions are met:
...
+- This script should be run from ncn-m001 when rebuilding ncn-m002 or ncn-m003.
```
Comment (deantroe) [inline operations/node_management/Rebuild_NCNs/Rebuild_NCNs.md:71]:
```suggestion
- This script should be run from `ncn-m001` when rebuilding `ncn-m002` or `ncn-m003`.
```

File: operations/iuf/stages/management_nodes_rollout.md [hunk 10]
```diff
@@ -26,22 +28,89 @@ The following arguments are most often used with the `management-nodes-rollout`
...
+
+## Manually rebuild NCNs with specific image and CFS configuration
+
+1. For all managment node types, first, get the image and CFS configuration created during `prepare-images` and `update-cfs-config` stages. Follow the instructions in [prepare-images](prepare_images.md#artifacts-created) to get the artifacts for `management-node-images`. For the image with the `configuration_group_name` matching
+the node type that is desired to be manually rebuilt (Management_Master, Management_Storage, Management_Worker), get the values for `final_image_id` and `configuration`.
+
+1. Next, follow the node rebuild instructions per node type.
+- Rebuild worker nodes
+
+    Worker nodes are automatically rebuilt using argo workflows. If rebuilding multiple worker nodes at once, see [this page](../../node_management/Rebuild_NCNs/Rebuild_NCNs.md#restrictions) for restrictions.
```
Comment (deantroe) [inline operations/iuf/stages/management_nodes_rollout.md:83]:
```suggestion
    Worker nodes are automatically rebuilt using Argo workflows. If rebuilding multiple worker nodes at once, see [this page](../../node_management/Rebuild_NCNs/Rebuild_NCNs.md#restrictions) for restrictions.
```

File: operations/iuf/stages/management_nodes_rollout.md [hunk 11]
```diff
@@ -26,22 +28,89 @@ The following arguments are most often used with the `management-nodes-rollout`
...
+
+In the case where it is necessary to start an entirely new worker rebuild workflow after a previous worker rebuild workflow failed, then the failed workflow must be deleted from Kubernetes. **Warning** that deleting a workflow will delete information about the state of that workflow and the steps that have been completed.
+
+To delete a failed Argo workflow, complete the following steps.
+
+1. Get the name of the failed workflow. All worker rebuild workflows start with `ncn-lifecycle-rebuild`. The name of the worker rebuild workflow can be found in the argo UI or by searching workflows in kubernetes with the following command.
```
Comment (deantroe) [inline operations/iuf/stages/management_nodes_rollout.md:104]:
```suggestion
1. Get the name of the failed workflow. All worker rebuild workflows start with `ncn-lifecycle-rebuild`. The name of the worker rebuild workflow can be found in the Argo UI or by searching workflows in Kubernetes with the following command.
```

File: operations/iuf/stages/management_nodes_rollout.md [hunk 12]
```diff
@@ -26,22 +28,89 @@ The following arguments are most often used with the `management-nodes-rollout`
...
+
+(`ncn-m#`) List failed worker rebuild workflows.
+```bash
+kubectl get Workflows -n argo | grep 'ncn-lifecycle-rebuild' | grep 'Fail'
```
Comment (deantroe) [inline operations/iuf/stages/management_nodes_rollout.md:108]:
```suggestion
kubectl get workflows -n argo | grep 'ncn-lifecycle-rebuild' | grep 'Fail'
```

File: operations/iuf/stages/management_nodes_rollout.md [hunk 13]
```diff
@@ -26,22 +28,89 @@ The following arguments are most often used with the `management-nodes-rollout`
...
+```
+
+1. (`ncn-m#`) Delete the failed workflow.
 ```bash
-iuf -a admin-230127 run --limit-management-rollout Management_Worker -r management-nodes-rollout
+kubectl delete Workflows -n argo FAILED_WORKFLOW
```
Comment (deantroe) [inline operations/iuf/stages/management_nodes_rollout.md:113]:
```suggestion
kubectl delete workflows -n argo FAILED_WORKFLOW
```

File: operations/iuf/stages/management_nodes_rollout.md [hunk 14]
```diff
@@ -1,19 +1,27 @@
...
-
-**`NOTE`** `management-nodes-rollout` currently does not rebuild management NCN storage nodes or `ncn-m001`. These nodes must be rebuilt using non-IUF methods described in the appropriate sections of the CSM documentation.
-
-- [Stage 1 - Ceph image upgrade](#../../upgrade/Stage_1.md) describes the rebuild process for NCN storage nodes
-- [Stage 2.3 - `ncn-m001` upgrade](#../../upgrade/Stage_2.md#stage-23---ncn-m001-upgrade) describes the rebuild process for `ncn-m001`
+The `management-nodes-rollout` stage performs a controlled update of management NCNs by configuring them with a new CFS configuration and/or rebuilding or upgrading them to a new image.
+A rebuild is a reboot operation that clears the persistent OverlayFS file system on that node.
+An upgrade is similar to a rebuild, but an upgrade does not wipe all information off of a node which is important
+specifically for NCN master and NCN storage nodes.
+An node image upgrade is only used during `management-nodes-rollout` if CSM is being upgraded.
```
Comment (deantroe) [inline operations/iuf/stages/management_nodes_rollout.md:7]: I think this will be misleading. On worker nodes, both COS and SHS currently add content to the base NCN worker image, and that modified image is what gets booted.  So CSM could remain at the same version, but we could take a patch for COS or SHS that results in an image "upgrade".   Maybe we should just remove this sentence?

File: operations/iuf/stages/management_nodes_rollout.md [hunk 15]
```diff
@@ -26,22 +34,105 @@ The following arguments are most often used with the `management-nodes-rollout`
...
+The `-cmrp` parameter selects the percentage of worker nodes that the worker node rebuild should coordinate rebuilding at one time.
+For example, if there are 15 worker nodes and `-cmrp 33` is specified, then 5 worker nodes will be rebuilt at once and with as much parallelization as possible given the state of the system.
 
-When NCN management nodes are being rebuilt, an additional Argo workflow will execute in addition to the standard IUF Argo workflows. It will include the string `ncn-lifecycle-rebuild` in its name.
+`-limit-management-rollout Management_Master` only needs to be specified when performing a CSM upgrade. This will upgrade `ncn-m002` and `ncn-m003` serially. This should be done before NCN worker nodes are upgraded.
+If not performing a CSM upgrade, then NCN master nodes should not be upgraded and should be configured with the CFS configuration created during the [update-cfs-config](../stages/update_cfs_config.md) stage.
 
-## Example
+## Examples
```
Comment (deantroe) [inline operations/iuf/stages/management_nodes_rollout.md:58]: Wondering if Examples should be at the bottom of the page, so the reader gets all of the information first before getting to the examples.

File: operations/iuf/stages/management_nodes_rollout.md [hunk 16]
```diff
@@ -26,22 +34,105 @@ The following arguments are most often used with the `management-nodes-rollout`
...
+
+1. `ncn-m002` will be upgraded
+1. `ncn-m003` will be upgraded
+1. Worker nodes `ncn-w001,ncn-w002,ncn-w003` will be upgraded
+1. Worker nodes `ncn-w005,ncn-w006,ncn-w007` will be upgraded
+1. Worker nodes `ncn-w008,ncn-w009,ncn-w010` will be upgraded
+
+## Manually upgrade or rebuild NCN worker node with specific image and CFS configuration
+
+The upgrade and rebuild procedure for NCN worker nodes is the same. This section applies to both NCN worker node upgrades and NCN worker node rebuilds.
```
Comment (deantroe) [inline operations/iuf/stages/management_nodes_rollout.md:93]:
```suggestion
The upgrade and rebuild procedures for NCN worker nodes are identical. This section applies to both NCN worker node upgrades and NCN worker node rebuilds.
```

File: operations/iuf/stages/management_nodes_rollout.md [hunk 17]
```diff
@@ -26,22 +34,105 @@ The following arguments are most often used with the `management-nodes-rollout`
...
+The words 'rebuild' and 'upgrade' are exchangeable in this section.
+
+1. Get the image and CFS configuration created during `prepare-images` and `update-cfs-config` stages.
+Follow the instructions in [prepare-images](prepare_images.md#artifacts-created) to get the artifacts for `management-node-images`. For the image with the `configuration_group_name` matching
+`Management_Worker`, get the values for `final_image_id` and `configuration`.
+
+1. Upgrade/rebuild worker nodes
+
+    Worker nodes are automatically rebuilt using Argo workflows. If rebuilding multiple worker nodes at once, see [this page](../../node_management/Rebuild_NCNs/Rebuild_NCNs.md#restrictions) for restrictions.
```
Comment (deantroe) [inline operations/iuf/stages/management_nodes_rollout.md:102]: I'd combine the above lines, so perhaps:
```
Upgrade/rebuild the worker nodes. Worker nodes are automatically rebuilt using Argo workflows. If rebuilding multiple worker nodes at once, see [this page](../../node_management/Rebuild_NCNs/Rebuild_NCNs.md#restrictions) for restrictions.
```

File: operations/iuf/stages/management_nodes_rollout.md [hunk 18]
```diff
@@ -26,22 +34,105 @@ The following arguments are most often used with the `management-nodes-rollout`
...
+    (`ncn-m001#`) Rebuild worker node.
+
+    ```bash
+    /usr/share/doc/csm/upgrade/scripts/upgrade/ncn-upgrade-worker-storage-nodes.sh ncn-w001 --image-id <final_image_id> --desired-cfs-conf <configuration>
+    ```
+
+## Action needed if a worker rebuild fails
+
+In general, worker node rebuilds should complete successfully before starting another rebuild.
+The node can get into a bad state if it has been partially rebuilt and then it is attempted to restart the rebuild on that same node.
```
Comment (deantroe) [inline operations/iuf/stages/management_nodes_rollout.md:113]:
```suggestion
A worker node can get into a bad state if it has been partially rebuilt and an attempt is made to restart the rebuild on that same node.
```

File: operations/iuf/stages/management_nodes_rollout.md [hunk 19]
```diff
@@ -26,22 +34,105 @@ The following arguments are most often used with the `management-nodes-rollout`
...
+In order to prevent this from happening, it is not possible to start another worker node rebuild if there is an incomplete worker node rebuild workflow.
```
Comment (deantroe) [inline operations/iuf/stages/management_nodes_rollout.md:114]:
```suggestion
As a result, it is not possible to start another worker node rebuild if there is an existing incomplete worker node rebuild workflow.
```

Comment (deantroe) [reply operations/iuf/stages/management_nodes_rollout.md:114]: If you like this, then the following sentence needs to be deleted as I sucked it into this sentence.

File: operations/iuf/stages/management_nodes_rollout.md [hunk 20]
```diff
@@ -26,22 +34,105 @@ The following arguments are most often used with the `management-nodes-rollout`
...
+Incomplete meaning it has stopped before successfully completing the full workflow. If there is an incomplete workflow and it is attempted to start another worker rebuild workflow,
+the first, incomplete worker rebuild workflow will continue and no new workflow will be created.
```
Comment (deantroe) [inline operations/iuf/stages/management_nodes_rollout.md:116]:
```
If an incomplete workflow exists and an attempt is made to start another worker rebuild workflow,
the original incomplete worker rebuild workflow will continue and no new workflow will be created.
```

File: operations/iuf/stages/management_nodes_rollout.md [hunk 21]
```diff
@@ -26,22 +34,105 @@ The following arguments are most often used with the `management-nodes-rollout`
...
+
+In the case where it is necessary to start an entirely new worker rebuild workflow after a previous worker rebuild workflow failed, then the failed workflow must be deleted from Kubernetes.
```
Comment (deantroe) [inline operations/iuf/stages/management_nodes_rollout.md:118]:
```suggestion
If it is necessary to start an entirely new worker rebuild workflow after a previous worker rebuild workflow failed, the failed workflow must be deleted from Kubernetes first.
```

File: operations/iuf/stages/management_nodes_rollout.md [hunk 22]
```diff
@@ -26,22 +34,105 @@ The following arguments are most often used with the `management-nodes-rollout`
...
+**Warning** that deleting a workflow will delete information about the state of that workflow and the steps that have been completed.
```
Comment (deantroe) [inline operations/iuf/stages/management_nodes_rollout.md:119]:
```suggestion
**WARNING** Deleting a workflow will delete information about the state of that workflow and the steps that have been completed.
```

File: operations/iuf/workflows/upgrade_all_products.md [hunk 23]
```diff
@@ -392,20 +392,250 @@ Once this step has completed:
...
+[Artifacts created](../stages/prepare_images.md#artifacts-created) section for details on how to query the images and [update-cfs-config](../stages/update_cfs_config.md) section for details about how the CFS configuration is updated.
 
 **`NOTE`** Additional arguments are available to control the behavior of the `management-nodes-rollout` stage, for example `--limit-management-rollout` and `-cmrp`. See the
 [`management-nodes-rollout` stage documentation](../stages/management_nodes_rollout.md) for details and adjust the examples below if necessary.
 
-**`NOTE`** The `management-nodes-rollout` stage creates additional separate Argo workflows when rebuilding NCN worker nodes. The Argo workflow names will include the string `ncn-lifecycle-rebuild`. If monitoring progress with the Argo UI,
-remember to include these workflows.
+**Important** There is a different procedure for  `management-nodes-rollout` depending on whether or not CSM is being upgraded.
```
Comment (deantroe) [inline operations/iuf/workflows/upgrade_all_products.md:400]:
```suggestion
**IMPORTANT** There is a different procedure for  `management-nodes-rollout` depending on whether or not CSM is being upgraded.
```

File: operations/iuf/workflows/upgrade_all_products.md [hunk 24]
```diff
@@ -392,20 +392,250 @@ Once this step has completed:
...
+The two procedures differ in the handling of NCN storage nodes and NCN master nodes. If CSM is not being upgraded, then NCN storage nodes and NCN master nodes will not be upgraded and will be updated by the CFS configuration
+created in [update-cfs-config](../stages/update_cfs_config.md).
+If CSM is being upgraded, the NCN storage nodes and NCN master nodes will be upgraded.
+Both procedures use the same steps for rebuilding/upgrading NCN worker nodes.
+Select **one** of the following procedures based on whether or not CSM is being upgraded.
+
+- [Management-nodes-rollout with CSM upgrade](#651-management-nodes-rollout-with-csm-upgrade)
```
Comment (deantroe) [inline operations/iuf/workflows/upgrade_all_products.md:407]: I think this section (and the one referenced in the line right below this one) is really about the master and storage nodes only, right?  If so, I'd reword as
```
NCN master and storage nodes with CSM upgrade scenario
```

That way, the flow is a bit more obvious/consistent, e.g the table of contents at the top will look like this:
```
6.5 Execute the IUF `management-nodes-rollout` stage
- 6.5.1 NCN master and strorage nodes with CSM upgrade
- 6.5.2 NCN master and storage nodes without CSM upgrade
- 6.5.3 NCN worker nodes
```

Or maybe it should even be:
```
6.5 Execute the IUF `management-nodes-rollout` stage
- 6.5.1 NCN master and strorage nodes
    - 6.5.1.1 with CSM upgrade
    - 6.5.1.2 without CSM upgrade
- 6.5.2 NCN worker nodes
```

File: operations/iuf/workflows/upgrade_all_products.md [hunk 25]
```diff
@@ -392,20 +392,250 @@ Once this step has completed:
...
+- [Management-nodes-rollout with no CSM upgrade](#652-management-nodes-rollout-with-no-csm-upgrade)
+
+#### 6.5.1 Management-nodes-rollout with CSM upgrade
+
+All management nodes will need to be upgraded to a new image because CSM is being upgraded. NCN master nodes, excluding `ncn-m001`, and NCN worker nodes can be upgraded with IUF.
+NCN storage nodes and `ncn-m001` will be upgraded with manual commands.
+This section describes how to test a new image and CFS configuration on a single "canary node" for NCN master nodes and NCN worker nodes first before rolling it out to the other NCN master nodes and NCN worker nodes.
```
Comment (deantroe) [inline operations/iuf/workflows/upgrade_all_products.md:414]: This should no longer refer to worker nodes as that is in a separate section now.

File: operations/iuf/workflows/upgrade_all_products.md [hunk 26]
```diff
@@ -392,20 +392,250 @@ Once this step has completed:
...
+Follow the steps below to upgrade all management nodes.
+
+1. Refer to the "Install and Upgrade Framework" section of each individual product's installation documentation to determine if any special actions need to be performed outside of IUF for the `management-nodes-rollout` stage.
+
+1. Get the image-id and CFS configuration created during `prepare-images` and `update-cfs-config` stages.
+Follow the instructions in [prepare-images](../stages/prepare_images.md#artifacts-created) to get the artifacts for `management-node-images`. For the images with the `configuration_group_name` matching
+`Management_Master` and `Management_Storage`, get the values for `final_image_id` and `configuration`.
+These values will be needed when upgrading NCN storage nodes and `ncn-m001`.
```
Comment (deantroe) [inline operations/iuf/workflows/upgrade_all_products.md:422]:
```suggestion
These values will be needed when upgrading NCN storage nodes and `ncn-m001` in the following steps.
```

File: operations/iuf/workflows/upgrade_all_products.md [hunk 27]
```diff
@@ -392,20 +392,250 @@ Once this step has completed:
...
+
+            Expected output:
+
+              ```bash
+              All components updated successfully.
+              ```
+
+    1. Set the image in BSS on all storage nodes by following the [Update NCN boot parameters](../../configuration_management/Management_Node_Image_Customization.md#4-update-ncn-boot-parameters)
+    section of [Management node image customizations](../../configuration_management/Management_Node_Image_Customization.md).
+    Set `IMS_RESULTANT_IMAGE_ID` variable to the `final_image_id` for `Management_Storage` found in the second step.
```
Comment (deantroe) [inline operations/iuf/workflows/upgrade_all_products.md:456]:
```suggestion
    Set the `IMS_RESULTANT_IMAGE_ID` variable to the `final_image_id` value for `Management_Storage` found in step 2 above.
```

File: operations/iuf/workflows/upgrade_all_products.md [hunk 28]
```diff
@@ -392,20 +392,250 @@ Once this step has completed:
...
+
+    1. (`ncn-m#`) Upgrade one NCN storage node.
+
+        **NOTE** This creates an additional, separate Argo workflow for rebuilding a NCN storage node. The Argo workflow name will include the string `ncn-lifecycle-rebuild`. If monitoring progress with the Argo UI, remember to include these workflows.
+
+        ```bash
+        /usr/share/doc/csm/upgrade/scripts/upgrade/ncn-upgrade-worker-storage-nodes.sh ncn-s001 --upgrade
+        ```
+
+    1. Verify that the storage node booted and is configured correctly. The CFS configuration can be checked will the command below using the `xname` of the node that was upgraded.
```
Comment (deantroe) [inline operations/iuf/workflows/upgrade_all_products.md:466]:
```suggestion
    1. Verify that the storage node booted and is configured correctly. The CFS configuration can be verified with the command below using the `xname` of the node that was upgraded.
```

File: operations/iuf/workflows/upgrade_all_products.md [hunk 29]
```diff
@@ -392,20 +392,250 @@ Once this step has completed:
...
+
+        ```bash
+        XNAME=x3000c0s13b0n0
+        cray cfs components describe $XNAME
+        ```
+
+        It is expected that `configurationStatus` is `configured`. If it is `pending`, then wait for the status to be `configured`.
```
Comment (deantroe) [inline operations/iuf/workflows/upgrade_all_products.md:473]:
```suggestion
        The desired value for `configurationStatus` is `configured`. If it is `pending`, then wait for the status to change to `configured`.
```

File: operations/iuf/workflows/upgrade_all_products.md [hunk 30]
```diff
@@ -392,20 +392,250 @@ Once this step has completed:
...
+        kubectl label nodes "ncn-m003" --overwrite iuf-prevent-rollout-
+        ```
+
+        (`ncn-m001#`) Verify the IUF node labels are present on the desired node.
+
+        ```bash
+        kubectl get nodes --show-labels | grep iuf-prevent-rollout
+        ```
+
+    1. Invoke `iuf run` with `-r` to execute the [`management-nodes-rollout`](../stages/management_nodes_rollout.md) stage on `ncn-m003`. This will rebuild the `ncn-m002` with the new CFS configuration and image built in
```
Comment (deantroe) [inline operations/iuf/workflows/upgrade_all_products.md:525]:
```suggestion
    1. Invoke `iuf run` with `-r` to execute the [`management-nodes-rollout`](../stages/management_nodes_rollout.md) stage on `ncn-m003`. This will rebuild the `ncn-m003` with the new CFS configuration and image built in
```

File: operations/iuf/workflows/upgrade_all_products.md [hunk 31]
```diff
@@ -392,20 +392,250 @@ Once this step has completed:
...
+    previous steps of the workflow.
+
+        (`ncn-m001#`) Execute the `management-nodes-rollout` stage with `ncn-m002`.
```
Comment (deantroe) [inline operations/iuf/workflows/upgrade_all_products.md:528]:
```suggestion
        (`ncn-m001#`) Execute the `management-nodes-rollout` stage with `ncn-m003`.
```

File: operations/iuf/workflows/upgrade_all_products.md [hunk 32]
```diff
@@ -392,20 +392,250 @@ Once this step has completed:
...
+
+            ```bash
+            CFS_CONFIG_NAME=configuration
+            ```
+
+        1. (`ncn-m#`) Get all NCN storage node xnames.
+
+            ```bash
+            XNAMES=$(cray hsm state components list --role Management --subrole Storage --type Node --format json | jq -r '.Components | map(.ID) | join(",")')
+            echo $XNAMES
```
Comment (deantroe) [inline operations/iuf/workflows/upgrade_all_products.md:438]: Mitch told me to document all variables like ```"${XNAMES}"```.  This change applies here and in other places below where $XNAMES and $XNAME are referred to.

File: operations/iuf/workflows/upgrade_all_products.md [hunk 33]
```diff
@@ -392,20 +392,250 @@ Once this step has completed:
...
+
+              ```bash
+              All components updated successfully.
+              ```
+
+    1. Set the image in BSS on `ncn-m001` by following the [Update NCN boot parameters](../../configuration_management/Management_Node_Image_Customization.md#4-update-ncn-boot-parameters)
+    section of [Management node image customizations](../../configuration_management/Management_Node_Image_Customization.md).
+    Set `IMS_RESULTANT_IMAGE_ID` variable to the `final_image_id` for `Management_Master` found in the second step.
+
+    1. (`ncn-m002#`) Upgrade `ncn-m001`. It is **important** to execute this from **`ncn-m002`**.
```
Comment (deantroe) [inline operations/iuf/workflows/upgrade_all_products.md:570]:
```suggestion
    1. (`ncn-m002#`) Upgrade `ncn-m001`. This **must** be executed on **`ncn-m002`**.
```

File: operations/iuf/workflows/upgrade_all_products.md [hunk 34]
```diff
@@ -392,20 +392,250 @@ Once this step has completed:
...
+stages create a new image and CFS configuration for NCN master nodes and storage nodes. The CFS configuration layers ensure the non-CSM product content is applied correctly for both
+image customization and node personalization scenarios. As a result, the administrator
+can update NCN master and storage nodes using CFS configuration only.
+Follow the following steps to complete the `management-nodes-rollout` stage.
+
+1. Rebuild the NCN worker nodes. Follow the procedure in section [6.5.3 NCN worker nodes](#653-ncn-worker-nodes). Then return to this procedure to complete the next step.
+
+1. Configure NCN master and NCN storage nodes.
+
+    1. (`ncn-m#`) Get the XNAMES for all NCN master and NCN storage nodes in a comma separated list.
```
Comment (deantroe) [inline operations/iuf/workflows/upgrade_all_products.md:591]:
```suggestion
    1. (`ncn-m#`) Get the xnames for all NCN master and NCN storage nodes in a comma separated list.
```

File: operations/iuf/workflows/upgrade_all_products.md (hunk 25 above)
Comment (leliasen-hpe) [reply operations/iuf/workflows/upgrade_all_products.md:414]: I am choosing to leave this as is. I think this is the overview of the section for upgrading all management nodes, including worker nodes. In my opinion it is not harmful to leave. It can be changed later if needed.
